### PR TITLE
Validate bond slaves have no mac duplicates

### PR DESCRIPTION
**Pull request by mmirecki**
This PR ensures that bond creation does not fail when slaves with duplicate macs are to be assigned to the bond.
Such a situation can arise when balance-alb bonds are created multiple times over. When such a bond is removed, the slaves are given back to the pool of free interfaces (for example vf's in case of sriov). If such interfaces with duplicate macs are used for another bond in balance-alb mode, the bond creation will fail due to duplicate macs.
In this PR we are detecting such a situation, and reset the duplicate mac to a random one on one of the conflicting interfaces.

### CODE REVIEW

**Comment by Eoghan1232** [comment]
@mmirecki will review today, can you look into the failed unit test?

**Comment by mmirecki** [comment]
> @mmirecki will review today, can you look into the failed unit test?

Looks like a random failure so set the mac address, with the error being:
`Failed to attached links to bond, error: Failed to handle duplicated macs on link slaves, error: cannot assign requested address`

I modified the PR to attempt to change the mac a few times (10) before acknowledging that an error has actually occurred.

**Comment by Eoghan1232** [comment]
@mmirecki can you run a go mod tidy?

ran locally and there was a diff:
`golang.org/x/sys v0.0.0-20210809222454-d867a43fc93e // indirect`

File: bond/bond_test.go [hunk 1]
```diff
@@ -85,7 +87,7 @@ var _ = Describe("tuning plugin", func() {
 			ContainerID: "dummy",
 			Netns:       podNS.Path(),
 			IfName:      IfName,
-			StdinData:   []byte(fmt.Sprintf(Config, "0.3.1")),
+			StdinData:   []byte(fmt.Sprintf(Config,  "0.3.1", ActiveBackupMode)),
```
Comment (Eoghan1232) [inline bond/bond_test.go:90]: remove empty spacing (covered by `go fmt`)

File: bond/bond_test.go [hunk 2]
```diff
@@ -231,4 +233,40 @@ var _ = Describe("tuning plugin", func() {
 			Expect(err).NotTo(HaveOccurred())
 		}
 	})
+
+	It("verifies the plugin copes with duplicated macs in  mode", func() {
```
Comment (Eoghan1232) [inline bond/bond_test.go:237]: balance-tlb mode ?

**Comment by Eoghan1232** [comment]
LGTM.
@eoghanlawless can you review?

**Comment by mmirecki** [comment]
> LGTM after those comment above are resolved. @eoghanlawless can you review?

Fixed

File: bond/util/validation.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+func updateDuplicateMac(link netlink.Link, netNsHandle *netlink.Handle, macsInUse []string) (string, error) {
+	var newMac string
+	var err error
+	// Setting of mac sometimes fails for various reasons. Retry a few times before failing.
```
Comment (eoghanlawless) [inline bond/util/validation.go:37]: This doesn't seem like a good way of getting around a failure.

What if it fails more than 10 times?

What are the various reasons and can we fix them?

File: bond/util/validation.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
+package util
```
Comment (eoghanlawless) [inline bond/util/validation.go:1]: Should this file be named `mac_validation.go` or will this file be used for other validation methods in the future?

**Review by eoghanlawless — CHANGES_REQUESTED** [review]
The loop to fix the failing test seems like the wrong approach.

If there's a good reason for it, add a more verbose comment.

Other than that everything looks great.

File: bond/util/validation.go (hunk 3 above)
Comment (mmirecki) [reply bond/util/validation.go:37]: Seems like it fails for specific macs. I tried to repeat this with just one mac, but then it fails repeatedly. Choosing a different mac seems to help.

File: bond/util/validation.go (hunk 4 above)
Comment (mmirecki) [reply bond/util/validation.go:1]: I hoped to get more of the validation code out of bond.go, for example the MTU one. Just did not want to polute this commit with unrelated changes.

File: bond/util/validation.go (hunk 3 above)
Comment (eoghanlawless) [reply bond/util/validation.go:37]: What are the specific macs, could we filter them out when generating the random macs?
Or at the very least, check if the mac fails and repeat indefinitely with a timeout, add better logging and create an issue to look at this in the future.

Comment (mmirecki) [reply bond/util/validation.go]: Limiting this to local/unicast seems to help.

**Review by Eoghan1232 — APPROVED** [review]
LGTM.

File: bond/util/validation.go (hunk 3 above)
Comment (eoghanlawless) [reply bond/util/validation.go]: LGTM

**Review by eoghanlawless — APPROVED** [review]

**Comment by mmirecki** [comment]
@Eoghan1232 @eoghanlawless could you please merge? Thanks

**Pull request merged**